src/main/webapp/shopping-results.js

```diff
@@ -0,0 +1,129 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * Builds the Shopping Results Page UI by integrating product results from 
+ * Google Shopping into the webpage. 
+ */
+async function buildShoppingResultsUI() {
+  // Make a GET request to '/search-shopping-results' to scrape the Google Shopping 
+  // query search results. The request returns the complete HTML of the SERP, stored
+  // into {@code shoppingSearchResultsPage}.
+
+  const response = await fetch('/search-shopping-results').catch(handleError);
+
+  if (!response.ok) {
+    return Promise.reject(response);
+  }
+
+  const shoppingSearchResultsPage = await response.text();
+
+  // Get the array of product containers from SERP - each has the class "u30d4".
+  const productElementsHTML = $('.u30d4', shoppingSearchResultsPage);
+
+  for (let i = 0; i < productElementsHTML.length - 1; i++) {
+    const currentProductHTML = productElementsHTML[i];
+
+    // Get info about the current product by extracting from the HTML element.
+
+    const productImageLink = $('.oR27Gd > img', currentProductHTML).attr('src');
+
+    let productLink = $('.rgHvZc > a', currentProductHTML).attr('href');
+
+    // Fix product link - if the URL starts with '/url?q=', the URL redirection will not work.
+    const wrongStartOfLink = '/url?q=';
+    // Therefore delete the start if this is the case, for the redirection to successfully work.
+    if (productLink.substring(0, wrongStartOfLink.length) == wrongStartOfLink) {
+      productLink = productLink.substring(wrongStartOfLink.length);
+    }
+
+    // Get the title as HTML instead of text, in order to keep the <b> tags.
+    const productTitle = $('.rgHvZc > a', currentProductHTML).html();
+
+    const productPrice = $('.dD8iuc > .HRLxBb', currentProductHTML).text()
```

Comment:
Nit: semicolon missing
```suggestion
    const productPrice = $('.dD8iuc > .HRLxBb', currentProductHTML).text();
```